### PR TITLE
Enable merge_group trigger for merge queues

### PR DESCRIPTION
**Pull request by tobiasdiez**
### 📚 Description

Two days ago [github announced the public beta of merge queues](https://github.blog/changelog/2023-02-08-pull-request-merge-queue-public-beta/). The idea is pretty similar the current release/merge process (as far as I understand it): Given you want to merge three PRs into `develop`, you add them to the merge queue and github will create three temporary branches:
- PR1
- PR1 + PR2 merged
- PR1 + PR2 + PR3 merged

It then runs all checks with the `merge_group` trigger on each branch. Once all checks for all these temporary branches are successful, the whole group is merged into `develop`. If there are merge conflicts, say between PR3 and PR1, or the build for the PR1+2+3 branch fails, then PR3 is removed from the merge queue and only the first two PRs are merged into `develop`. 

In this way, one could never end up in a situation where the checks are not passing in the develop branch, without requiring devs to always merge the latest develop branch themselves. (E.g., we will no longer hit the situation that PR2 introduces new linter rules which are not fullfiled by PR3 but both PRs are merged into develop).

References:
- https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue
- https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#merge_group

---

In this PR we add the merge_group trigger on the current PR-workflows, which enables to run these in the merge queue as well.

---

Since this effectively implements the current version of "Volker's temporary private develop branch", I was wondering if this would be eneough to give more people rights to merge PRs. What else would be necessary for such a step? @vbraun 

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by dimpase** [comment]
why is `master` branch mentioned there?

**Comment by tobiasdiez** [comment]
Github will create temporary branches that will trigger the `push` event. This change restricts the push trigger to the only two branches that we currently have and ignores these temporary branches. Currently the `push` trigger is equivalent to `push: -master -develop`.

**Comment by dimpase** [comment]
maybe I'm mistreading what the patch does, but appearnce of `master` right next to `develop` makes me think you want to do this queue on `master` as well as on `develop`

**Comment by dimpase** [comment]
> change restricts the push trigger to the only two branches that we currently have

but why do we need it on `master`? I'd say we don't...

**Comment by tobiasdiez** [comment]
> maybe I'm mistreading what the patch does, but appearnce of `master` right next to `develop` makes me think you want to do this queue on `master` as well as on `develop`

This is a bit of a stupid yml syntax. The indented "branches" only apply to "push" and not to "merge_queue".



> > change restricts the push trigger to the only two branches that we currently have
> 
> but why do we need it on `master`? I'd say we don't...

I was not sure about this change, since the role of the master branch is not clear to me. If it is indeed only an old copy of the develop branch and does not receive any updates, then sure we don't need to run these workflows on it (but then why is there a master branch in the first place?). If the master branch may be updated (hotfixes?) than you probably want to verify that the new commits adhere at least to these basic tests.

**Comment by mkoeppe** [comment]
The `master` branch is fast-forwarded to stable release tags made on the `develop` branch.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -2,7 +2,11 @@ name: Build & Test
 
 on:
   pull_request:
+  merge_group:
   push:
+    branches:
+      - master
+      - develop
```
Comment (mkoeppe) [inline .github/workflows/build.yml:9]: This change to `push` is not good. Developers may want to see these workflows run on their forks when they push branches, before they open pull requests.

Comment (tobiasdiez) [reply .github/workflows/build.yml:9]: What is your use case for this?

You can always run the workflow manually in your fork if you want, or create a PR in your fork or a draft PR against the main repo. I think this selective choice is better than running every push to a PR from a fork with actions enabled through two passes.

Comment (mkoeppe) [reply .github/workflows/build.yml:9]: The main use cases are experimental code that is not ready for a pull request but should undergo continuous testing; and long-running branches that merge various changes for use in a class or a research project that involves several people; cf. https://groups.google.com/g/sage-devel/c/ayOL8_bzOfk/m/QUrnDONMCAAJ

In any case, this change has nothing to do with merge queues.

Comment (tobiasdiez) [reply .github/workflows/build.yml:9]: So you are saying that the majority of the runs in https://github.com/mkoeppe/sage/actions/ were desired and helpful?
The use cases you describe can easily be covered by creating a draft PR against the main repo, or a PR in your fork. 

> In any case, this change has nothing to do with merge queues.

It does, at described in https://github.com/sagemath/sage/pull/35062#issuecomment-1426054216

Comment (mkoeppe) [reply .github/workflows/build.yml:9]: So what are the temp branches named?

**Comment by tobiasdiez** [comment]
@vbraun Did you already found the time to look into merge queues? In its current form this PR doesn't replace your buildbots but it ensures that PR's don't introduce changes that make other PRs fail (when merged at the same time). This happened a few times in the last months and is pretty annoying since then the github workflow fails for every new PR.

**Comment by tobiasdiez** [comment]
@dimpase since you were interested in the merge queue feature. 
My proposal would be to merge this PR now. Then we can test it by merging some test PRs into some test branch in the main repo (for which I can activate the merge queue independent of the main repo). Once these tests are successful, we can approach @vbraun and discuss how to best integrate the merge queues with his release manager scripts. What do you think?

**Comment by dimpase** [comment]
could it rather have a manual trigger, for the time being?

**Comment by dimpase** [comment]
master branch ought to be removed here, IMHO

**Comment by tobiasdiez** [comment]
One can only trigger a merge queue by merging PRs. There is no other way.

(The master branch there is just for completeness - it looks like its no longer used, but officially its still there.)

**Comment by dimpase** [comment]
does this mean that PRs are merged automatically? 
Into a temporary branch?

So then they are still mergeable in the usual way?

**Comment by tobiasdiez** [comment]
For the existing PRs nothing changes. But I can configure it in github ui that all PRs that target a new 'test' branch have to go through merge queues

**Comment by dimpase** [comment]
what does it mean for PR to " target" a test branch?

**Comment by dimpase** [comment]
Oh, right, of course. So, suppose this happens, and the PR gets more or less automatically merged into 'test'.
How is it going to reach 'develop' branch?

**Comment by tobiasdiez** [comment]
Never ;-) The point of the 'test' branch would be to merge stupid test PRs (some nonsensical changes in the readme say, some others that break the build etc) in order to test that the merge queue works. Once this is established, we can approach @vbraun and ask that real PRs are merged into the develop branch using the merge queue.

**Review by dimpase — APPROVED** [review]
OK

**Comment by mkoeppe** [comment]
Do these experiments to get familiar with merge queues really have to be done in this repo? I'd suggest not.

**Comment by dimpase** [comment]
Why not? It's not breaking anything

**Comment by tobiasdiez** [comment]
> Pull request merge queues are available in any public repository owned by an organization

So you cannot try it out in a fork.

Thanks @dimpase for your review!

**Comment by tobiasdiez** [comment]
I've now enabled the merge queue feature. You can try it out by creating a PR against https://github.com/sagemath/sage/tree/test_merge_queue. For example, https://github.com/sagemath/sage/pull/36432.

Currently, only the linter workflow has to pass for the merge to succeed.

**Pull request merged**